crates/wit-parser/src/ast.rs

```diff
@@ -1226,15 +1330,58 @@ impl SourceMap {
                     self.require_f32_f64,
                 )
                 .with_context(|| format!("failed to tokenize path: {}", src.path.display()))?;
-                let ast = Ast::parse(&mut tokens)?;
-                resolver.push(ast).with_context(|| {
-                    format!("failed to start resolving path: {}", src.path.display())
-                })?;
+                match Ast::parse(&mut tokens)? {
+                    Ast::ExplicitPackages(pkgs) => {
+                        for pkg in pkgs {
+                            let mut resolver = Resolver::default();
+                            let pkg_name = pkg.package_id.package_name();
+                            let ingested = resolver.push_then_resolve(pkg).with_context(|| {
+                                format!("failed to start resolving path: {}", src.path.display())
+                            })?;
+
+                            match explicit_pkg_names.get(&pkg_name) {
+                                Some(_) => bail!(
+                                    "colliding explicit package names, multiple packages named `{}`",
+                                    pkg_name
+                                ),
+                                None => explicit_pkg_names.insert(pkg_name),
+                            };
+
+                            if let Some(unresolved) = ingested {
+                                ret_packages.push(unresolved);
+                            }
+                        }
+                    }
+                    Ast::PartialImplicitPackage(partial) => {
+                        if !explicit_pkg_names.is_empty() {
+                            bail!("WIT files cannot mix top-level explicit `package` declarations with other declaration kinds 2");
+                        }
+
+                        implicit_resolver.push_partial(partial).with_context(|| {
+                            format!("failed to start resolving path: {}", src.path.display())
+                        })?;
+                    }
+                }
+            }
+
+            if let Some(pkg) = implicit_resolver.resolve()? {
```

Comment:
If you'd be ok with something slightly wordier here I'm a bit worried about this mixing of partial-and-explicit resolvers and using the wrong resolver by accident. Could this function perhaps have:

```rust
enum PackageResolver {
    Unknown,
    Explicit,
    Implicit(Resolver),
}
```

where it'd start as `Unknown` and transition to other states once a package was seen. That way there's no way for the `ExplicitPackages` side to accidentally use the implicit resolver since it won't ever match on it. Additionally here this method wouldn't be called for explicit packages, only implicit ones, so the `resolve()` method wouldn't need to return an `Option`

Reply:
Done. I went with a slightly different structure for what I termed `ResolverKind`, which also tracks `Explicit` variants' parsed packages in the ADT as well.